### PR TITLE
Add an example to demonstrate custom extern

### PR DESCRIPTION
**Pull request by qobilidop**
I spent some time learning how to do this myself as a result of [this discussion](https://github.com/p4lang/p4c/pull/3502). I thought it might be useful to add an example here to help others. This example roughly follows @jafingerhut's [suggestion from a previous discussion](https://github.com/p4lang/behavioral-model/issues/697#issuecomment-565759070).

### CODE REVIEW

File: examples/custom_extern/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+Adding interface veth0 as port 0
+Adding interface veth2 as port 1
+```
+
+In another terminal, we can do the testing by sending a test packet to `simple_switch`, and check the packet received:
+```console
+$ make test
+```
+
+If everything works as expected, you shall see:
```
Comment (antoninbas) [inline examples/custom_extern/README.md:33]: s/shall/should

File: examples/custom_extern/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+Packet received:
+###[ Ethernet ]### 
+  dst       = ff:ff:ff:ff:ff:ff
+  src       = 42:01:c0:a8:3f:48
+  type      = 0x9000
+###[ Raw ]### 
+     load      = '\\xbd\x11\x01\x00'
+```
+
+When you are done with testing:
```
Comment (antoninbas) [inline examples/custom_extern/README.md:57]:
```suggestion
When you are done with testing, tun the following after killing the `simple_switch` process:
```

File: examples/custom_extern/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+$ make stop
+$ make clean
+```
+
+## Reference
+
+- Previous discussion: https://github.com/p4lang/behavioral-model/issues/697
+- Relevant section in the P4 Language Specification: https://p4.org/p4-spec/docs/P4-16-v1.2.3.html#sec-external-units
+- Relevant testing code in bmv2: https://github.com/p4lang/behavioral-model/blob/main/tests/test_extern.cpp
+- Another example: https://github.com/engjefersonsantiago/p4-programs/tree/master/examples/div_by_n
```
Comment (antoninbas) [inline examples/custom_extern/README.md:68]: could you make them markdown links?

**Review by antoninbas — COMMENTED** [review]
this is a great contribution, thanks!

could you add Apache copyright header to new source files (cpp, p4, py) as appropriate?

File: examples/custom_extern/Makefile [hunk 4]
```diff
@@ -0,0 +1,32 @@
+P4_PROGRAM = example.json
```
Comment (antoninbas) [inline examples/custom_extern/Makefile:1]: that's a bit confusing to call the json file `P4_PROGRAM`
maybe you should use `P4_PROGRAM := example`

File: examples/custom_extern/README.md (hunk 1 above)
Comment (qobilidop) [reply examples/custom_extern/README.md]: Done.

File: examples/custom_extern/README.md (hunk 3 above)
Comment (qobilidop) [reply examples/custom_extern/README.md]: Done.

File: examples/custom_extern/Makefile (hunk 4 above)
Comment (qobilidop) [reply examples/custom_extern/Makefile]: Done.

File: examples/custom_extern/example.p4 [hunk 5]
```diff
@@ -1,3 +1,18 @@
+/* Copyright 2013-present Barefoot Networks, Inc.
```
Comment (antoninbas) [inline examples/custom_extern/example.p4:1]: you need to change this line to
```
/* Copyright 2022 <WHATEVER IS APPROPRIATE IN YOUR CASE>
```
or potentially:
```
/* Copyright 2022 P4lang Authors
```

same in other files

Comment (qobilidop) [reply examples/custom_extern/example.p4]: Fixed.

**Review by antoninbas — APPROVED** [review]

**Comment by aneeshnotfound** [comment]
I am running into segmentation faults when trying to modify P4 variables inside a C++ function within simple_switch_grpc. While passing variables as const or by value works for reading, arithmetic, and logging, passing them as non-const references leads to garbage memory allocation and crashes. I need to ensure that changes made to these variables in C++ properly reflect back in the P4 code. Does simple_switch_grpc support mutable P4 variable references in external C++ functions, and if so, what is the correct way to pass them?

**Comment by jafingerhut** [comment]
> I am running into segmentation faults when trying to modify P4 variables inside a C++ function within simple_switch_grpc. While passing variables as const or by value works for reading, arithmetic, and logging, passing them as non-const references leads to garbage memory allocation and crashes. I need to ensure that changes made to these variables in C++ properly reflect back in the P4 code. Does simple_switch_grpc support mutable P4 variable references in external C++ functions, and if so, what is the correct way to pass them?

I would recommend creating a new issue on this repository, and link to this closed PR for reference.

**Pull request merged**